Expand filename_parser to handle time series and stage 3 filenames

bourque opened the pull request:
This PR modifies `utils.filename_parser()` to be able to handle filenames for time series observations (e.g. `jw00733003001_02101_00002-seg001_nrs1_rate.fits`) and stage 3 products (e.g. `jw80600-o009_t001-epoch1_miri_f1130w_i2d.fits`). 

Corresponding tests have also been added to `test_utils.py`, and `test_utils.py` was modified to utilize `pytest.mark.parametrize`.

Closes #237

pep8speaks commented:
Hello @bourque, Thank you for submitting the Pull Request !

 - In the file [`jwql/tests/test_utils.py`](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py), following are the PEP8 issues :

> [Line 28:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L28): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 29:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L29): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 40:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L40): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 41:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L41): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 52:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L52): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 53:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L53): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 63:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L63): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 64:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L64): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 72:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L72): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 73:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L73): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 82:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L82): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 83:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L83): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 90:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L90): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 91:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L91): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 93:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L93): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 94:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L94): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 95:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L95): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 96:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L96): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 97:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L97): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 98:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L98): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 99:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L99): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 100:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L100): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 101:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L101): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 103:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L103): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 104:1](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L104): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 106:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L106): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 107:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L107): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 108:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L108): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 109:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L109): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 110:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L110): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 111:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L111): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 112:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L112): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 113:5](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L113): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 148:101](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/tests/test_utils.py#L148): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (156 > 100 characters)

 - In the file [`jwql/utils/utils.py`](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py), following are the PEP8 issues :

> [Line 38:17](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L38): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 39:17](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L39): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 102:16](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L102): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 102:19](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L102): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\w'
> [Line 102:21](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L102): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 102:22](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L102): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 102:23](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L102): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 102:23](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L102): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 102:26](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L102): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 103:22](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L103): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 103:22](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L103): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\w'
> [Line 116:16](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L116): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 116:16](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L116): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 116:21](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L116): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 116:21](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L116): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 116:30](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L116): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\w'
> [Line 125:21](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L125): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 125:22](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L125): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 125:26](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L125): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 125:26](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L125): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 126:28](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L126): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 126:31](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L126): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\w'
> [Line 136:16](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L136): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 136:19](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L136): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\w'
> [Line 136:20](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L136): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\w'
> [Line 136:21](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L136): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 136:22](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L136): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 136:22](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L136): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 136:23](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L136): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 136:23](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L136): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 137:20](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L137): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 137:21](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L137): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\d'
> [Line 163:101](https://github.com/spacetelescope/jwql/blob/154c643d4b955a5446c575869498aead5a866c75/jwql/utils/utils.py#L163): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (164 > 100 characters)

If you have not done so, please consult the [`jwql` Style Guide](https://github.com/spacetelescope/jwql/blob/master/style_guide/style_guide.md)

bourque commented:
I don't think there is anything in @pep8speaks to really worry about.  A lot of these are because of code alignment, which makes the code easier to read.

bourque commented:
@laurenmarietta 

> looking at the JDox page, it says that the source ID in association files can be up to 5 digits. When that is being parsed in lines 118 and 128, the ID is restricted to 3 digits (which is the limit for the target ID). I worry that could end in an error?

Good catch!  We can probably can get around this by adding another `filename_type` wherein a `source_id` is used instead of a `target_id`.  I'll see if I can implement that. 

> Which makes me think: the best way to test this ultimately would be to run the filename_parser on all the filenames we currently have in the filesystem (which we determined were all valid names, right?) and make sure it can handle them all. Probably not a good idea as the filesystem scales up, but might be good to do earlier on so we're sure we've covered all the bases. I'd be happy to try and do that if you like.

This is a good idea!  Feel free to open a separate issue/PR for this.  We could write a unit test to test if **all** filenames pass, instead of having a unit test that tests a given file individually -- using the latter could inflate our test suite by a lot, if that makes sense.

laurenmarietta commented:
Got it - I'm on it!

bourque commented:
@laurenmarietta My latest commit (https://github.com/spacetelescope/jwql/pull/248/commits/a35108c472e32cd0fca78eec808a922be83339d1) should address the concerns about `source_id`.  I also moved some code around to avoid some redundant code, and added declarations for raw strings to avoid some `DeprecationWarnings` (see [relevant StackOverflow](https://github.com/spacetelescope/jwql/pull/248/commits/a35108c472e32cd0fca78eec808a922be83339d1))

laurenmarietta commented:
I'm writing the test for the whole filesystem (only 333 out of ~10,000 are failing!), and there are some specific filetypes that are failing: 
- jw90001001002_02101_00001_nis_**o001**_crf.fits
- jw00307001001_02101_00001_nis_**a3001**_**crfints**.fits
- jw00322002007_02101_00001_nrcalong_**c1001**_crf.fits

So it looks like 
1) `crfints` should be added to the `FILE_SUFFIX_TYPES` list, and 
2) there is another common filetype where the AC_ID is after the _instrument_ rather than after the _proposal number_. 

Thought I'd share so you could include those changes in this PR?

laurenmarietta commented on jwql/utils/utils.py:
```diff
@@ -109,31 +94,108 @@ def filename_parser(filename):
+    stage_3_target_id = \
+        r"jw" \
+        r"(?P<program_id>\d{5})"\
+        r"-(?P<ac_id>(o|c|a|r)\d{3})"\
```
This also just made me realize that AC_IDs with `c`, `a`, or `r` can have 4 digits where the regex string is currently looking for 3. Also on lines 126, 135, and 145.

bourque replied:
@laurenmarietta 😱 Nice catch.  I'll try to get around this one too.  Perhaps there is a regex option where i can say "3 or 4".

bourque commented:
@laurenmarietta 

> there is another common filetype where the AC_ID is after the instrument rather than after the proposal number.

I suggest that we hold off on coding a workaround for this one.  I would like to ask the powers that be over in the archive if this is expected, since it goes directly against what they have documented.  Or perhaps the filesystem that we are using to test needs to be updated.

Johannes-Sahlmann commented:
Examples of FGS GS acquisition and fine-guide file names in MAST:
jw00788003001_gs-acq1_2019160140055_uncal.fits
jw00788003001_gs-fg_2019160140451_uncal.fits
Note that these are of PRODUCT CATEGORY=AUXILIARY (as opposed to SCIENCE)

bourque commented:
Thanks @Johannes-Sahlmann 

Some additional documentation I've been made aware of thanks to Kim Duprie:

https://innerspace.stsci.edu/pages/viewpage.action?pageId=94092600#SummaryofAssociationFileNamingConvention

https://innerspace.stsci.edu/display/SCSB/JWST+Science+Data+Products

bourque commented:
I'm going to move this back to a `WIP` until I can make sense of all of these different filename types.

bourque commented:
On second thought, I think we should go ahead and merge these changes and worry about future changes in a separate PR, especially considering this PR is holding up the progress of #255 

Looking at all of the possibilities described in [this document](https://innerspace.stsci.edu/pages/viewpage.action?pageId=94092600), I think we are going to have to make some rather significant changes, and we can worry about those whenever we update our filesystem to a newer copy (still waiting to hear back from Mary Romelfanger about this).  That way, we should have some of these filetypes available to us for testing.

@laurenmarietta Were there any other changes necessary for this PR?
